Automattic/wp-calypso
# Eliminate moment locales from server build

When building the Calypso server, both web and desktop variants, the server bundle includes all moment.js locale definitions. That's because the `MomentProvider` component contains a dynamic `import( `moment/locale/${ currentLocale }` )` statement in its locale-loading code.

But we don't use the `MomentProvider` in the server code at all, and the server never does any localized date formatting. The provider is bundled only because the _context consumer_ in React UI code imports the `components/localized-moment/context` module, which drags the provider into the bundle with all its dependencies.

We want to continue to use the context consumer in the SSR-ed React UI, but use the default context value that's used when there is no provider rendered. I.e., the EN locale.

This PR reorganizes the `components/localized-moment` folder in such a way that importing the consumer never imports the provider as well. Therefore, bundling the provider only in the browser client bundle that really uses it.

Also, it adds webpack `IgnorePlugin` definitions to ignore dynamic `require` calls inside moment.js itself. The moment.js code contains a function like:
```js
function loadLocale( name ) {
  require( './locale/' + name );
}
```

**How to test:**
Verify that the server bundle, after `NODE_ENV=production yarn run build-server`, is significantly smaller.

Before:
```
$ ls -l build/
total 22304
-rw-r--r--  1 jsnajdr  staff  4056334 Oct  1 10:40 server.js
-rw-r--r--  1 jsnajdr  staff  5827737 Oct  1 10:40 server.js.map
-rw-r--r--  1 jsnajdr  staff    18856 Oct  1 10:40 server.lib-analytics-signup.js
-rw-r--r--  1 jsnajdr  staff    32743 Oct  1 10:40 server.lib-analytics-signup.js.map
-rw-r--r--  1 jsnajdr  staff     5053 Oct  1 10:40 server.moment-locale-af.js
-rw-r--r--  1 jsnajdr  staff     7583 Oct  1 10:40 server.moment-locale-af.js.map
-rw-r--r--  1 jsnajdr  staff   654263 Oct  1 10:40 server.vendors~moment-locale-af.js
-rw-r--r--  1 jsnajdr  staff   806672 Oct  1 10:40 server.vendors~moment-locale-af.js.map
```

After:
```
$ ls -l build/
total 19344
-rw-r--r--  1 jsnajdr  staff  4043005 Oct  1 10:42 server.js
-rw-r--r--  1 jsnajdr  staff  5803474 Oct  1 10:42 server.js.map
-rw-r--r--  1 jsnajdr  staff    18856 Oct  1 10:42 server.lib-analytics-signup.js
-rw-r--r--  1 jsnajdr  staff    32743 Oct  1 10:42 server.lib-analytics-signup.js.map
```

You should see a similar improvement in the desktop build (`yarn run build-desktop:server`):

Before:
```
$ ls -l desktop/build/
total 13008
-rw-r--r--  1 jsnajdr  staff  6650606 Oct  1 10:58 desktop.js
-rw-r--r--  1 jsnajdr  staff     7760 Oct  1 10:58 desktop.lib-analytics-signup.js
```

After:
```
$ ls -l desktop/build/
total 12560
-rw-r--r--  1 jsnajdr  staff  6419190 Oct  1 10:59 desktop.js
-rw-r--r--  1 jsnajdr  staff     7760 Oct  1 10:59 desktop.lib-analytics-signup.js
```

The Desktop app bundles all `node_modules` into the `desktop.js` bundle, so the chunk layout is different, but the size improvement should be there, too.
